Userland/Libraries/LibWeb/Bindings/WindowObject.cpp

```diff
@@ -437,4 +447,114 @@ JS_DEFINE_NATIVE_FUNCTION(WindowObject::inner_height_getter)
     return JS::Value(impl->inner_height());
 }
 
+JS_DEFINE_NATIVE_GETTER(WindowObject::scroll_x_getter)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return JS::Value(0.0);
+    return JS::Value(static_cast<double>(impl->document().page()->top_level_browsing_context().viewport_scroll_offset().x()));
+}
+
+JS_DEFINE_NATIVE_GETTER(WindowObject::scroll_y_getter)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return JS::Value(0.0);
+    return JS::Value(static_cast<double>(impl->document().page()->top_level_browsing_context().viewport_scroll_offset().y()));
+}
+
+enum class ScrollBehavior {
+    Auto,
+    Smooth
+};
+
+static void perform_a_scroll(Page& page, double x, double y, ScrollBehavior)
+{
+    // FIXME: Stop any existing smooth-scrolls
+    // FIXME: Implement smooth-scroll
+    page.client().page_did_request_scroll_to({ x, y });
+}
+
+// https://drafts.csswg.org/cssom-view/#dom-window-scroll
+JS_DEFINE_NATIVE_FUNCTION(WindowObject::scroll)
+{
+    auto* impl = impl_from(vm, global_object);
+    if (!impl)
+        return {};
+    if (!impl->document().page())
+        return {};
+    auto& page = *impl->document().page();
+
+    auto viewport_rect = page.top_level_browsing_context().viewport_rect();
+    JS::Value x_value = JS::Value(viewport_rect.x());
+    JS::Value y_value = JS::Value(viewport_rect.y());
+    String behavior_string = "auto";
+
+    if (vm.argument_count() == 1) {
+        auto& options = vm.argument(0).as_object();
+        if (auto left = options.get("left"); !left.is_undefined())
+            x_value = left;
+        if (auto top = options.get("top"); !top.is_undefined())
+            y_value = top;
+        behavior_string = options.get("behavior").to_string_without_side_effects();
```

Comment:
Putting this in a separate comment so it's not hidden in an edit:
Testing in another engine, behavior_string is the first thing to be fetched, so this should come before `left` and `top`.